hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -1778,21 +1780,30 @@ def __init__(self, x, type, indices=Indices(), aggregations=LinkedList(Aggregati
             self._set_field(f, expr)
 
     def _set_field(self, key, value):
-        self._fields[key] = value
-        if key not in self.__dict__:
-            self.__dict__[key] = value
+        if key not in self._fields:
+            # Avoid using hasattr on self. Each new field added will fall through to __getattr__,
+            # which has to build a nice error message.
+            if key in self.__dict__ or hasattr(super(), key):
+                self._warn_on_shadowed_name.add(key)
+            else:
+                self.__dict__[key] = value
+            self._fields[key] = value
 
     def _get_field(self, item):
         if item in self._fields:
             return self._fields[item]
         else:
             raise KeyError(get_nice_field_error(self, item))
 
+    def __getattribute__(self, item):
+        if item in super().__getattribute__('_warn_on_shadowed_name'):
```

Comment:
This avoids infinite recursion, right? And the super's getattribute is able to see the subclasses basic fields?

Reply:
Yes, and yes. The super's getattribute is just what would have been called if we didn't override it (unless there's some special-method magic I don't understand).